[FEATURE] [PERF] WebGL kernel and gradient for unsortedSegmentSum

jgartman opened the pull request:
This PR makes a couple of improvements to the `unsortedSegmentSum` function that was added as part of the gradient for `gather` #981 .

1. Adds a WebGL kernel for `unsortedSegmentSum`.
1. Adds a gradient.
1. Moves `unsortedSegmentSum` and its tests into their own files.  This may not be necessary but there's a couple of reasons I thought it was a good idea.  The output of `unsortedSegmentSum` can have equal or larger size compared to it's input so I don't think it really belongs with the reduction ops.  Also there are other segment ops that we may get requests to support someday so it would be nice to have a place for them.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-core/1051)
<!-- Reviewable:end -->

nsthorat commented:
Nice work! Few comments, one pressing one is about "axis" which doesn't exist in TensorFlow Py.

---

Review status: 6 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/kernels/webgl/segment_gpu.ts, line 11 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLFaQ9LtUpMBlv6qa:-LEFLFaQ9LtUpMBlv6qb:bq92rk8) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/kernels/webgl/segment_gpu.ts#L11)):*
> ```XML
>   constructor(
>       segOpInfo: SegOpInfo,
>       segOpType: 'unsortedSegmentSum',
> ```

remove trailing comma

---

*[src/ops/segment_ops.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLApN4BxNN7078woW:-LEFLApN4BxNN7078woX:b-4lizs5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops.ts#L1)):*
> ```XML
> import {doc} from '../doc';
> ```

add a license to this file

---

*[src/ops/segment_ops.ts, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFNA4TAeiEXgrO8goB:-LEFNA4TAeiEXgrO8goC:be6fg6r) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops.ts#L35)):*
> ```XML
>   @operation
>   static unsortedSegmentSum<T extends Tensor>(
>       x: T, segmentIds: Tensor1D, numSegments: number, axis = 0): T {
> ```

TensorFlow py doesn't have axis so we won't be able to implement this in node, is there a reason we have this here?

---

*[src/ops/segment_ops_test.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLAAY9g8NBHrkdGkv:-LEFLAAY9g8NBHrkdGkw:b-4lizs5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops_test.ts#L1)):*
> ```XML
> import * as tf from '../index';
> ```

add a license to this file

---

*[src/ops/segment_util.ts, line 2 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFL7OK6-fMCDqfgUHG:-LEFL7OK6-fMCDqfgUHH:b-4lizs5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_util.ts#L2)):*
> ```XML
> import {nearestDivisor} from '../util';
> import {PARALLELIZE_THRESHOLD} from './reduce_util';
> ```

add a license to this file

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-:-LEFPCBb5Gf-yjMC2XCp:bld0uw2)*
<!-- Sent from Reviewable.io -->

jgartman commented:
Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/kernels/webgl/segment_gpu.ts, line 11 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLFaQ9LtUpMBlv6qa:-LEaXIAn7g99oigfd1un:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/kernels/webgl/segment_gpu.ts#L11)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove trailing comma
</blockquote></details>

Done.

---

*[src/ops/segment_ops.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLApN4BxNN7078woW:-LEaXKoo3qi_3xYvO-DH:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops.ts#L1)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

add a license to this file
</blockquote></details>

Done.

---

*[src/ops/segment_ops.ts, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFNA4TAeiEXgrO8goB:-LEaXLlA2hqa5xgzmL-D:b-rp4tzo) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops.ts#L35)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

TensorFlow py doesn't have axis so we won't be able to implement this in node, is there a reason we have this here?
</blockquote></details>

Not really, `unsortedSegmentSum` was originally written as a private helper function used by the gradient for `gather` and the axis parameter was there because it made the gradient easier to calculate.  It should have been removed before it became part of the public API.  I removed it, reworked the gradient for `gather` and wrote a few more tests to make sure the `gather` gradient still works as expected.

---

*[src/ops/segment_ops_test.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFLAAY9g8NBHrkdGkv:-LEaY58nEXfycOQyA7VM:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_ops_test.ts#L1)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

add a license to this file
</blockquote></details>

Done.

---

*[src/ops/segment_util.ts, line 2 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEFL7OK6-fMCDqfgUHG:-LEaY6X10UeL_ZX-RV-H:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e5ce35658185d33022a1290c9297dd04efc83b2f/src/ops/segment_util.ts#L2)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

add a license to this file
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1051)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

Nice work Josh, this is impressive & thorough! One small comment about docs then I will merge this.

---

Review status: 1 of 15 files reviewed at latest revision, all discussions resolved.

---

*[src/ops/segment_ops.ts, line 39 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-LEf2qWB69zXL3XEVW2G:-LEf2qWC27eRYmCZhgAy:bmwhlwy) ([raw file](https://github.com/tensorflow/tfjs-core/blob/ec0a16eae264fb32af0eb84b50742ccd1f28c998/src/ops/segment_ops.ts#L39)):*
> ```XML
>    *
>    * x.unsortedSegmentSum(indices, numSegments).print() //or
>    * tf.unsortedSegmentSum(x, indices, numSegments)
> ```

since this is on its own line, it will also be run as code when this is in the doc generator, can you make sure theres a comment in front of it?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1051#-:-LEf38kPB-xxTwWv9KYV:b-w7r87b)*
<!-- Sent from Reviewable.io -->
